keps/sig-api-machinery/4346-informer-metrics/README.md

```diff
@@ -0,0 +1,973 @@
+
+Each reflector metrics contains 3 counter, 4 summary and 1 gauge.
+```
+type reflectorMetrics struct {
+	numberOfLists       CounterMetric
+	listDuration        SummaryMetric
+	numberOfItemsInList SummaryMetric
+
+	numberOfWatches      CounterMetric
+	numberOfShortWatches CounterMetric
+	watchDuration        SummaryMetric
+	numberOfItemsInWatch SummaryMetric
+
+	lastResourceVersion GaugeMetric
+}
+```
+According to kubernetes/kubernetes#73587, the memory leak is caused by summary. Every summary contains `AgeBuckets`of buckets which contains an array with 500 quantile.Sample. Reduce summary `ageBuckets` to mitigate memory usage. 
+
+The listDuration/numberOfItemsInList/watchDuration/numberOfItemsInWatch will not `Observe` very frequently. According to [prometheus client](https://github.com/prometheus/client_golang/commit/15c9ded5a3b7ae08886e51ba26a97270ca23fecd#diff-4c7dc681b10f8e28a3d6a8cfd115c37cc473e6fc9b0949bd6b41eaedbdead438R132), it is enough to set `ageBuckets` to 1.
+
+### Remove Metrics 
+When the informers and reflectors stopped, the reference metrics will be removed.
+
+Kube component-base metrics support to delete metrics by matching labels.
+
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, for Alpha try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+The data can be easily read from:
+https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+- `<package>`: `<date>` - `<test coverage>`
+
+##### Integration tests
```

Comment:
IIRC there was some tests for metrics and granularity of them @logicalhan @dgrisonnet can you help us here?

Reply:
We do have some precedence of integration tests for metrics such as these ones for the apiserver: https://github.com/kubernetes/kubernetes/blob/87fa400d9ddf4fa48ab32a7d745a3b60dc785e5d/test/integration/metrics/metrics_test.go as well as some utilities to write some https://github.com/kubernetes/kubernetes/blob/87fa400d9ddf4fa48ab32a7d745a3b60dc785e5d/staging/src/k8s.io/component-base/metrics/testutil/testutil.go.

Reply:
Add some tests  for metrics subsystem/label/granularity, PTAL

Reply:
> Add some tests for metrics subsystem/label/granularity, PTAL

forgot to push?